File: src/core/engine.ts
import { CategoryKey, Difficulty, Question, QuestionView, AnswerResult, SessionSummary } from '../types';
import { DIFFICULTY_XP, questionsForCategory } from '../data/questions';
import { checkNewAchievements } from '../data/achievements';
import { PlayerProfile } from './player';

export const QUESTIONS_PER_GAME = 10;
export const FIFTY_FIFTY_USES = 2;
export const SKIP_USES = 2;
export const STREAK_STEP = 0.1;
export const STREAK_CAP = 10;

export class QuizSession {
  profile: PlayerProfile;
  category: CategoryKey | 'mixed' | null;
  rng: () => number;
  questions: Question[];

  index = 0;
  answered = 0;
  correctCount = 0;
  wrongCount = 0;
  skippedCount = 0;
  streak = 0;
  bestStreak = 0;
  xpEarned = 0;
  levelUps: number[] = [];
  lifelines = { fifty: FIFTY_FIFTY_USES, skip: SKIP_USES };

  private currentOptions: string[] = [];
  private currentCorrect = 0;
  private currentRaw?: Question;
  private fiftyRemoved = new Set<number>();
  private fiftyUsedHere = false;
  private answeredCurrent = false;

  constructor(
    profile: PlayerProfile,
    category?: CategoryKey | 'mixed' | null,
    numQuestions = QUESTIONS_PER_GAME,
    rng: () => number = Math.random
  ) {
    this.profile = profile;
    this.category =
      category && category !== 'mixed'
        ? (category as CategoryKey)
        : null;
    this.rng = rng;

    const pool = questionsForCategory(this.category);
    this.questions = shuffle(pool, this.rng).slice(0, Math.max(1, numQuestions));
  }

  get total(): number {
    return this.questions.length;
  }

  get finished(): boolean {
    return this.index >= this.total;
  }

  get perfect(): boolean {
    return (
      this.wrongCount === 0 &&
      this.skippedCount === 0 &&
      this.correctCount === this.total
    );
  }

  currentQuestion(): QuestionView {
    const raw = this.questions[this.index];
    const order = shuffleIndices(4, this.rng);
    this.currentOptions = order.map((i) => raw.options[i]);
    this.currentCorrect = order.indexOf(raw.answer);
    this.currentRaw = raw;
    this.fiftyRemoved.clear();
    this.fiftyUsedHere = false;
    this.answeredCurrent = false;

    return {
      qid: raw.id,
      category: raw.category,
      difficulty: raw.difficulty as Difficulty,
      text: raw.question,
      options: [...this.currentOptions],
      position: this.index + 1,
      total: this.total,
    };
  }

  answer(optionIndex: number): AnswerResult {
    if (this.finished || this.answeredCurrent) {
      throw new Error('No active question to answer');
    }
    if (this.fiftyRemoved.has(optionIndex)) {
      throw new Error('Option was removed by 50:50');
    }

    this.answeredCurrent = true;
    const raw = this.currentRaw!;
    const isCorrect = optionIndex === this.currentCorrect;
    this.answered += 1;
    let xp = 0;
    const gainedLevels: number[] = [];

    if (isCorrect) {
      this.correctCount += 1;
      this.streak += 1;
      this.bestStreak = Math.max(this.bestStreak, this.streak);
      const base = DIFFICULTY_XP[raw.difficulty as Difficulty];
      const mult = 1 + STREAK_STEP * Math.min(this.streak - 1, STREAK_CAP);
      xp = Math.round(base * mult);
      const { oldLevel, newLevel } = this.profile.addXp(xp);
      this.xpEarned += xp;
      if (newLevel > oldLevel) {
        for (let l = oldLevel + 1; l <= newLevel; l += 1) {
          gainedLevels.push(l);
          this.levelUps.push(l);
        }
      }
    } else {
      this.wrongCount += 1;
      this.streak = 0;
    }

    this.profile.recordAnswer(raw.category, isCorrect);
    this.index += 1;

    return {
      correct: isCorrect,
      correctIndex: this.currentCorrect,
      xpGained: xp,
      fact: raw.fact,
      streak: this.streak,
      finished: this.finished,
      levelUps: gainedLevels,
    };
  }

  useFiftyFifty(): number[] {
    if (this.answeredCurrent || this.finished) {
      throw new Error('Cannot use 50:50 now');
    }
    if (this.fiftyUsedHere) {
      throw new Error('50:50 already used on this question');
    }
    if (this.lifelines.fifty <= 0) {
      throw new Error('No 50:50 lifelines left');
    }

    const wrong = Array.from({ length: 4 }, (_, i) => i).filter(
      (i) => i !== this.currentCorrect && !this.fiftyRemoved.has(i)
    );
    const removed = shuffle(wrong, this.rng).slice(0, 2).sort((a, b) => a - b);
    removed.forEach((i) => this.fiftyRemoved.add(i));
    this.fiftyUsedHere = true;
    this.lifelines.fifty -= 1;
    return removed;
  }

  useSkip(): void {
    if (this.answeredCurrent || this.finished) {
      throw new Error('Cannot skip now');
    }
    if (this.lifelines.skip <= 0) {
      throw new Error('No skip lifelines left');
    }
    this.lifelines.skip -= 1;
    this.skippedCount += 1;
    this.index += 1;
  }

  get lifelinesUsed(): number {
    return (
      FIFTY_FIFTY_USES -
      this.lifelines.fifty +
      (SKIP_USES - this.lifelines.skip)
    );
  }

  summary(): SessionSummary {
    return {
      category: this.category ?? 'mixed',
      answered: this.answered,
      correct: this.correctCount,
      wrong: this.wrongCount,
      skipped: this.skippedCount,
      total: this.total,
      bestStreak: this.bestStreak,
      xpEarned: this.xpEarned,
      perfect: this.perfect,
      lifelinesUsed: this.lifelinesUsed,
      levelUps: [...this.levelUps],
    };
  }

  finalize() {
    this.profile.recordSession(this.bestStreak, this.perfect);
    const newBadges = checkNewAchievements(this.profile.toJSON(), this.summary());
    newBadges.forEach((b) => this.profile.addAchievement(b.id));
    return newBadges;
  }
}

function shuffle<T>(arr: T[], rng: () => number): T[] {
  const copy = [...arr];
  for (let i = copy.length - 1; i > 0; i -= 1) {
    const j = Math.floor(rng() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

function shuffleIndices(n: number, rng: () => number): number[] {
  return shuffle(Array.from({ length: n }, (_, i) => i), rng);
}

File: gyaanquest/core/engine.py
"""Quiz session engine: question flow, scoring, streaks and lifelines."""
from __future__ import annotations

import random
from dataclasses import dataclass

from gyaanquest.data.questions import (
    CATEGORIES, DIFFICULTY_XP, questions_for_category,
)
from gyaanquest.core import achievements as ach

QUESTIONS_PER_GAME = 10
FIFTY_FIFTY_USES = 2
SKIP_USES = 2
# Streak multiplier: +10% XP per consecutive correct answer, capped at +100%.
STREAK_STEP = 0.10
STREAK_CAP = 10


@dataclass
class QuestionView:
    qid: str
    category: str
    difficulty: str
    text: str
    options: list          # shuffled
    position: int          # 1-based
    total: int


@dataclass
class AnswerResult:
    correct: bool
    correct_index: int
    xp_gained: int
    fact: str
    streak: int
    finished: bool
    level_ups: list        # levels gained by this answer, e.g. [4]


class QuizSession:
    def __init__(self, profile, category=None, num_questions=QUESTIONS_PER_GAME, rng=None):
        self.profile = profile
        self.category = category if category in CATEGORIES else None
        self.rng = rng or random.Random()

        pool = questions_for_category(self.category)
        self.rng.shuffle(pool)
        self.questions = pool[:max(1, num_questions)]

        self.index = 0
        self.answered = 0
        self.correct_count = 0
        self.wrong_count = 0
        self.skipped_count = 0
        self.streak = 0
        self.best_streak = 0
        self.xp_earned = 0
        self.level_ups = []
        self.lifelines = {"fifty": FIFTY_FIFTY_USES, "skip": SKIP_USES}
        self._fifty_removed = set()   # option indices removed on current question
        self._answered_current = False
        self._shuffled = []           # prepared questions cache

    # ---------------- question flow ----------------
    @property
    def total(self):
        return len(self.questions)

    @property
    def finished(self):
        return self.index >= self.total

    def current_question(self) -> QuestionView:
        raw = self.questions[self.index]
        order = list(range(4))
        self.rng.shuffle(order)
        self._current_options = [raw["options"][i] for i in order]
        self._current_correct = order.index(raw["answer"])
        self._current_raw = raw
        self._fifty_removed = set()
        self._fifty_used_here = False
        self._answered_current = False
        return QuestionView(
            qid=raw["id"],
            category=raw["category"],
            difficulty=raw["difficulty"],
            text=raw["question"],
            options=list(self._current_options),
            position=self.index + 1,
            total=self.total,
        )

    # ---------------- answering ----------------
    def answer(self, option_index: int) -> AnswerResult:
        if self.finished or self._answered_current:
            raise RuntimeError("No active question to answer")
        if option_index in self._fifty_removed:
            raise ValueError("Option was removed by 50:50")

        self._answered_current = True
        raw = self._current_raw
        is_correct = option_index == self._current_correct
        self.answered += 1
        xp = 0
        gained_levels = []

        if is_correct:
            self.correct_count += 1
            self.streak += 1
            self.best_streak = max(self.best_streak, self.streak)
            base = DIFFICULTY_XP[raw["difficulty"]]
            mult = 1.0 + STREAK_STEP * min(self.streak - 1, STREAK_CAP)
            xp = int(round(base * mult))
            old_level, new_level = self.profile.add_xp(xp)
            self.xp_earned += xp
            if new_level > old_level:
                gained_levels = list(range(old_level + 1, new_level + 1))
                self.level_ups.extend(gained_levels)
        else:
            self.wrong_count += 1
            self.streak = 0

        self.profile.record_answer(raw["category"], is_correct)
        self.index += 1

        return AnswerResult(
            correct=is_correct,
            correct_index=self._current_correct,
            xp_gained=xp,
            fact=raw["fact"],
            streak=self.streak,
            finished=self.finished,
            level_ups=gained_levels,
        )

    # ---------------- lifelines ----------------
    def use_fifty_fifty(self) -> list:
        """Remove two wrong options. Return the removed option indices."""
        if self._answered_current or self.finished:
            raise RuntimeError("Cannot use 50:50 now")
        if self._fifty_used_here:
            raise RuntimeError("50:50 already used on this question")
        if self.lifelines["fifty"] <= 0:
            raise RuntimeError("No 50:50 lifelines left")
        wrong = [i for i in range(4)
                 if i != self._current_correct and i not in self._fifty_removed]
        self.rng.shuffle(wrong)
        removed = sorted(wrong[:2])
        self._fifty_removed.update(removed)
        self._fifty_used_here = True
        self.lifelines["fifty"] -= 1
        return removed

    def use_skip(self) -> None:
        """Skip the current question without losing the streak."""
        if self._answered_current or self.finished:
            raise RuntimeError("Cannot skip now")
        if self.lifelines["skip"] <= 0:
            raise RuntimeError("No skip lifelines left")
        self.lifelines["skip"] -= 1
        self.skipped_count += 1
        self.index += 1

    @property
    def lifelines_used(self):
        return (FIFTY_FIFTY_USES - self.lifelines["fifty"]
                + SKIP_USES - self.lifelines["skip"])

    # ---------------- wrap up ----------------
    @property
    def perfect(self):
        return (self.wrong_count == 0 and self.skipped_count == 0
                and self.correct_count == self.total)

    def summary(self) -> dict:
        return {
            "category": self.category or "mixed",
            "answered": self.answered,
            "correct": self.correct_count,
            "wrong": self.wrong_count,
            "skipped": self.skipped_count,
            "total": self.total,
            "best_streak": self.best_streak,
            "xp_earned": self.xp_earned,
            "perfect": self.perfect,
            "lifelines_used": self.lifelines_used,
            "level_ups": list(self.level_ups),
        }

    def finalize(self) -> list:
        """Record the session on the profile; return newly unlocked achievements."""
        self.profile.record_session(self.best_streak, self.perfect)
        new_badges = ach.check_new(self.profile, self.summary())
        self.profile.achievements.extend(b.id for b in new_badges)
        return new_badges
